internal/worker_versioning_info.go

```diff
@@ -0,0 +1,310 @@
+	// It is part of [TaskQueueInfo].
+	TaskQueueVersionInfo struct {
+		// Task queue info per task type.
+		TypesInfo map[TaskQueueType]TaskQueueTypeInfo
+		// The category of tasks that may reach a versioned worker of a certain Build ID.
+		TaskReachability BuildIDTaskReachability
+	}
+
+	// TaskQueueInfo is the response to [Client.DescribeTaskQueueEnhanced].
+	TaskQueueInfo struct {
+		// Task queue information for each Build ID. Empty string as key value means unversioned.
+		VersionsInfo map[string]TaskQueueVersionInfo
+	}
+)
+
+func (o *DescribeTaskQueueEnhancedOptions) validateAndConvertToProto(namespace string) (*workflowservice.DescribeTaskQueueRequest, error) {
+	if namespace == "" {
+		return nil, errors.New("missing namespace argument")
+	}
+
+	if o.TaskQueue == "" {
+		return nil, errors.New("missing task queue field")
+	}
+
+	taskQueueTypes := make([]enumspb.TaskQueueType, len(o.TaskQueueTypes))
+	for i, t := range o.TaskQueueTypes {
+		taskQueueTypes[i] = taskQueueTypeToProto(t)
+	}
+
+	opt := &workflowservice.DescribeTaskQueueRequest{
+		Namespace: namespace,
+		TaskQueue: &taskqueuepb.TaskQueue{
+			// Sticky queues not supported
+			Name: o.TaskQueue,
+		},
+		ApiMode:                enumspb.DESCRIBE_TASK_QUEUE_MODE_ENHANCED,
+		Versions:               taskQueueVersionSelectionToProto(o.Versions),
+		TaskQueueTypes:         taskQueueTypes,
+		ReportPollers:          o.ReportPollers,
+		ReportTaskReachability: o.ReportTaskReachability,
+	}
+
+	return opt, nil
+}
+
+func workerVersionCapabilitiesFromResponse(response *common.WorkerVersionCapabilities) *WorkerVersionCapabilities {
+	if response == nil {
+		return nil
+	}
+
+	return &WorkerVersionCapabilities{
+		BuildID:       response.GetBuildId(),
+		UseVersioning: response.GetUseVersioning(),
+	}
+}
+
+func pollerInfoFromResponse(response *taskqueuepb.PollerInfo) PollerInfo {
+	if response == nil {
+		return PollerInfo{}
+	}
+
+	lastAccessTime := time.Time{}
+	if response.GetLastAccessTime() != nil {
+		lastAccessTime = response.GetLastAccessTime().AsTime()
+	}
+
+	return PollerInfo{
+		LastAccessTime:            lastAccessTime,
+		Identity:                  response.GetIdentity(),
+		RatePerSecond:             response.GetRatePerSecond(),
+		WorkerVersionCapabilities: workerVersionCapabilitiesFromResponse(response.GetWorkerVersionCapabilities()),
+	}
+}
+
+func taskQueueTypeInfoFromResponse(response *taskqueuepb.TaskQueueTypeInfo) TaskQueueTypeInfo {
+	if response == nil {
+		return TaskQueueTypeInfo{}
+	}
+
+	pollers := make([]PollerInfo, len(response.GetPollers()))
+	for i, pInfo := range response.GetPollers() {
+		pollers[i] = pollerInfoFromResponse(pInfo)
+	}
+
+	return TaskQueueTypeInfo{
+		Pollers: pollers,
+	}
+}
+
+func taskQueueVersionInfoFromResponse(response *taskqueuepb.TaskQueueVersionInfo) TaskQueueVersionInfo {
+	if response == nil {
+		return TaskQueueVersionInfo{}
+	}
+
+	typesInfo := make(map[TaskQueueType]TaskQueueTypeInfo)
```

Comment:
```suggestion
	typesInfo := make(map[TaskQueueType]TaskQueueTypeInfo, len(response.GetTypesInfo()))
```

Pedantic good practice on map creation (here and elsewhere), not required